src/dcrypto/v1/v1.go

```diff
@@ -0,0 +1,240 @@
+	return aesKey, hmacKey, nil
+}
+
+type hashReadWriter struct {
+	hash hash.Hash
+	done bool
+	sum  io.Reader
+}
+
+func (h *hashReadWriter) Write(p []byte) (int, error) {
+	if h.done {
+		return 0, errors.New("writing to hashReadWriter after read is not allowed")
+	}
+	return h.hash.Write(p)
+}
+
+func (h *hashReadWriter) Read(p []byte) (int, error) {
+	if !h.done {
+		h.done = true
+		h.sum = bytes.NewBuffer(h.hash.Sum(nil))
+	}
+	return h.sum.Read(p)
+}
+
+// NewEncryptReader returns an io.Reader wrapping the provided io.Reader.
+// It uses a user provided password and a random salt to derive keys.
+// If the key is provided interactively, it should be verified since there
+// is no recovery.
+func NewEncryptReader(r io.Reader, pass []byte) (io.Reader, error) {
+	salt, err := randBytes(saltSize)
+	if err != nil {
+		return nil, err
+	}
+	return newEncryptReader(r, pass, salt, scryptIter)
+}
+
+// newEncryptReader returns a encryptReader wrapping an io.Reader.
+// It uses a user provided password and the provided salt iterated the
+// provided number of times to derive keys.
+func newEncryptReader(r io.Reader, pass, salt []byte, iterations int) (io.Reader, error) {
+	aesKey, hmacKey, err := keys(pass, salt, iterations)
+	if err != nil {
+		return nil, err
+	}
+	b, err := aes.NewCipher(aesKey)
+	if err != nil {
+		return nil, err
+	}
+	h := hmac.New(hashFunc, hmacKey)
+	iv, err := randBytes(blockSize)
+	if err != nil {
+		return nil, err
+	}
+	hr := &hashReadWriter{hash: h}
+	sr := &cipher.StreamReader{R: r, S: cipher.NewCTR(b, iv)}
+	var header []byte
+	header = append(header, salt...)
+	header = append(header, iv...)
+	return io.MultiReader(io.TeeReader(io.MultiReader(bytes.NewBuffer(header), sr), hr), hr), nil
+}
+
+// decryptReader wraps a io.Reader decrypting its content.
+type decryptReader struct {
+	tmpFile *os.File
+	sReader *cipher.StreamReader
+}
+
+// NewDecryptReader creates an io.ReadCloser wrapping an io.Reader.
+// It has to read the entire io.Reader to disk using a temp file so that it can
+// hash the contents to verify that it is safe to decrypt.
+// If the file is athenticated, the DecryptReader will be returned and
+// the resulting bytes will be the plaintext.
+func NewDecryptReader(r io.Reader, pass []byte) (io.ReadCloser, error) {
+	return newDecryptReader(r, pass, scryptIter)
+}
+
+func newDecryptReader(r io.Reader, pass []byte, iterations int) (d *decryptReader, err error) {
+	salt := make([]byte, saltSize)
+	iv := make([]byte, blockSize)
+	mac := make([]byte, hmacSize)
+	_, err = io.ReadFull(r, salt)
+	if err != nil {
+		return nil, err
+	}
+	_, err = io.ReadFull(r, iv)
+	if err != nil {
+		return nil, err
+	}
+	aesKey, hmacKey, err := keys(pass, salt, iterations)
+	if err != nil {
+		return nil, err
+	}
+	// Start Verifying the HMAC of the message.
+	h := hmac.New(hashFunc, hmacKey)
+	h.Write(salt)
+	h.Write(iv)
+	dst, err := ioutil.TempFile("", "drive-encrypted-")
```

Comment:
We could write a wrapper for a TempFile and add a .Close() or .Done() method to it so that we can avoid having to deal with .Close() then os.Remove(...) everytime we need to use a temp file. Let me spin up for you a package.